pkg/kubectl/history_test.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func Test_HistoryViewerFor(t *testing.T) {
+	fakeClientset := &fake.Clientset{}
+	daemonSetKind := schema.GroupKind{Group: "apps", Kind: "DaemonSet"}
+	result, err := HistoryViewerFor(daemonSetKind, fakeClientset)
+	if err != nil {
+		t.Fatalf("error getting HistoryViewer for a DaemonSet: %v", err)
+	}
+
+	if _, ok := result.(*DaemonSetHistoryViewer); !ok {
+		t.Fatalf("unexpected output type (DaemonSetHistoryViewer was expected)")
+	}
+
+	statefulKind := schema.GroupKind{Group: "apps", Kind: "StatefulSet"}
+	result, err = HistoryViewerFor(statefulKind, fakeClientset)
+	if err != nil {
+		t.Fatalf("error getting HistoryViewer for a StatefulSet: %v", err)
+	}
+
+	if _, ok2 := result.(*StatefulSetHistoryViewer); !ok2 {
+		t.Fatalf("unexpected output type (StatefulSetHistoryViewer was expected)")
+	}
+
+	deploymentKind := schema.GroupKind{Group: "apps", Kind: "Deployment"}
+	result, err = HistoryViewerFor(deploymentKind, fakeClientset)
+	if err != nil {
+		t.Fatalf("error getting HistoryViewer for a Deployment: %v", err)
+	}
+
+	if _, ok3 := result.(*DeploymentHistoryViewer); !ok3 {
+		t.Fatalf("unexpected output type (DeploymentHistoryViewer was expected)")
+	}
```

Comment:
Suggest making this a table of test cases (ref https://github.com/golang/go/wiki/TableDrivenTests). Suggest adding a case when no history viewer is returned and an expected error is returned instead.

Reply:
> Suggest making this a table of test cases.

I was thinking about suggesting that before. But I realize type assertion may make it hard.

Reply:
Yes. Storing types in a table will not be straightforward.

Reply:
We can do something like:

```go
var expectedType reflect.Type
expectedType = reflect.TypeOf(&DaemonSetHistoryViewer{})
if reflect.TypeOf(result) == expectedType {
	t.Fatalf("unexpected output type, expected %v, got %v", expectedType, reflect.TypeOf(result))
}
```

This can be easily converted into tables.